File: vmidentity/ssoclients/restclient/idmclient/src/user_resource.c
/*
 * Copyright © 2012-2016 VMware, Inc.  All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the “License”); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an “AS IS” BASIS, without
 * warranties or conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

#include "includes.h"

static PCSTRING const TENANT_URI = "/idm/tenant";
static PCSTRING const TENANT_POST_URI = "/idm/post/tenant";

SSOERROR
IdmExternalUserCreate(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    const IDM_PRINCIPAL_DATA* pPrincipal,
    bool** ppCreated,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING resourceUri = NULL;
    bool* pCreated = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || IS_NULL_OR_EMPTY_STRING(tenant) || pPrincipal == NULL
        || ppCreated == NULL || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildResourceUri(
        pClient,
        TENANT_URI,
        tenant,
        "users",
        NULL,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        pPrincipal,
        (DataObjectToJsonFunc) IdmPrincipalDataToJson,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_POST,
        (JsonToDataObjectFunc) RestJsonToBooleanData,
        (void**) &pCreated,
        &pError);
    BAIL_ON_ERROR(e);

    *ppCreated = pCreated;

    // debug
    if (DEBUG)
    {
        RestDebugJsonObject(*ppCreated, (DataObjectToJsonFunc) RestBooleanDataToJson);
    }

    error:

    if (e != SSOERROR_NONE)
    {
        RestBooleanDataDelete(pCreated);
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(resourceUri);

    return e;
}

SSOERROR
IdmUserGet(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    PCSTRING name,
    PCSTRING domain,
    IDM_USER_DATA** ppUserReturn,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING upn = NULL;
    PSTRING resourceUri = NULL;
    IDM_USER_DATA* pUserReturn = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || IS_NULL_OR_EMPTY_STRING(tenant) || IS_NULL_OR_EMPTY_STRING(name)
        || IS_NULL_OR_EMPTY_STRING(domain) || ppUserReturn == NULL || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildUPN(name, domain, &upn);
    BAIL_ON_ERROR(e);

    e = RestBuildResourceUri(
        pClient,
        TENANT_POST_URI,
        tenant,
        "users",
        upn,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        NULL,
        NULL,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_POST,
        (JsonToDataObjectFunc) IdmJsonToUserData,
        (void**) &pUserReturn,
        &pError);
    BAIL_ON_ERROR(e);

    *ppUserReturn = pUserReturn;

    // debug
    if (DEBUG)
    {
        RestDebugJsonObject(*ppUserReturn, (DataObjectToJsonFunc) IdmUserDataToJson);
    }

    error:

    if (e != SSOERROR_NONE)
    {
        IdmUserDataDelete(pUserReturn);
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(upn);
    SSOStringFree(resourceUri);

    return e;
}

SSOERROR
IdmUserGetGroups(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    PCSTRING name,
    PCSTRING domain,
    bool nested,
    IDM_GROUP_ARRAY_DATA** ppGroupArrayReturn,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING upn = NULL;
    PSTRING resourceUri = NULL;
    IDM_GROUP_ARRAY_DATA* pGroupArrayReturn = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || IS_NULL_OR_EMPTY_STRING(tenant) || IS_NULL_OR_EMPTY_STRING(name)
        || IS_NULL_OR_EMPTY_STRING(domain) || ppGroupArrayReturn == NULL || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildUPN(name, domain, &upn);
    BAIL_ON_ERROR(e);

    e = RestBuildResourceUri(
        pClient,
        TENANT_POST_URI,
        tenant,
        "users",
        upn,
        "groups",
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestAppendQueryStringOnResourceUri("nested", nested ? "TRUE" : "FALSE", true, resourceUri, &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        NULL,
        NULL,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_POST,
        (JsonToDataObjectFunc) IdmJsonToGroupArrayData,
        (void**) &pGroupArrayReturn,
        &pError);
    BAIL_ON_ERROR(e);

    *ppGroupArrayReturn = pGroupArrayReturn;

    // debug
    if (DEBUG)
    {
        RestDebugJsonArray(*ppGroupArrayReturn, (DataObjectToJsonFunc) IdmGroupArrayDataToJson);
    }

    error:

    if (e != SSOERROR_NONE)
    {
        IdmGroupArrayDataDelete(pGroupArrayReturn);
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(upn);
    SSOStringFree(resourceUri);

    return e;
}

SSOERROR
IdmExternalUserDelete(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    PCSTRING name,
    PCSTRING domain,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING upn = NULL;
    PSTRING resourceUri = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || IS_NULL_OR_EMPTY_STRING(tenant) || IS_NULL_OR_EMPTY_STRING(name)
        || IS_NULL_OR_EMPTY_STRING(domain) || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildUPN(name, domain, &upn);
    BAIL_ON_ERROR(e);

    e = RestBuildResourceUri(
        pClient,
        TENANT_URI,
        tenant,
        "users",
        upn,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        NULL,
        NULL,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_DELETE,
        NULL,
        NULL,
        &pError);
    BAIL_ON_ERROR(e);

    error:

    if (e != SSOERROR_NONE)
    {
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(upn);
    SSOStringFree(resourceUri);

    return e;
}

File: vmidentity/rest/idm/client/src/main/java/com/vmware/identity/rest/idm/client/UserResource.java
/*
 *  Copyright (c) 2012-2015 VMware, Inc.  All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"); you may not
 *  use this file except in compliance with the License.  You may obtain a copy
 *  of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS, without
 *  warranties or conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the
 *  License for the specific language governing permissions and limitations
 *  under the License.
 */
package com.vmware.identity.rest.idm.client;

import static com.vmware.identity.rest.core.client.RequestExecutor.execute;
import static com.vmware.identity.rest.core.client.RequestExecutor.executeAndReturnList;
import static com.vmware.identity.rest.core.client.URIFactory.buildParameters;
import static com.vmware.identity.rest.core.client.URIFactory.buildURI;

import java.io.IOException;
import java.net.URI;
import java.util.List;
import java.util.Map;

import org.apache.http.HttpException;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.methods.HttpPost;

import com.vmware.identity.rest.core.client.BaseClient;
import com.vmware.identity.rest.core.client.ClientResource;
import com.vmware.identity.rest.core.client.RequestFactory;
import com.vmware.identity.rest.core.client.UPNUtil;
import com.vmware.identity.rest.core.client.exceptions.ClientException;
import com.vmware.identity.rest.core.client.exceptions.WebApplicationException;
import com.vmware.identity.rest.idm.data.GroupDTO;
import com.vmware.identity.rest.idm.data.UserDTO;

/**
 * The {@code UserResource} is effectively a container that gathers all of
 * the commands related to a tenant's users. These commands are all gathered
 * under the {@value #USER_URI} endpoint on the REST server.
 */
public class UserResource extends ClientResource {

    private static final String USER_POST_URI = "/idm/post/tenant/%s/users";
    private static final String USER_NAME_POST_URI = USER_POST_URI + "/%s";
    private static final String USER_NAME_GROUPS_POST_URI = USER_NAME_POST_URI + "/groups";

    public UserResource(BaseClient parent) {
        super(parent);
    }

    /**
     * Request a specific user.
     *
     * <p><b>Required Role:</b> {@code user}.
     *
     * @see UPNUtil#buildUPN(String, String)
     * @param tenant the tenant to retrieve the user from.
     * @param name the name of the user to request.
     * @param domain the domain of the user to request.
     * @return the requested user.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public UserDTO get(String tenant, String name, String domain) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        String upn = UPNUtil.buildUPN(name, domain);
        URI uri = buildURI(parent.getHostRetriever(), USER_NAME_POST_URI, tenant, upn);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken());
        return execute(parent.getClient(), post, UserDTO.class);
    }

    /**
     * Request the list of groups that a user is a member of.
     *
     * <p><b>Required Role:</b> {@code user}.
     *
     * @param tenant the name of the tenant containing the user.
     * @param name the name of the user.
     * @param domain the domain of the user.
     * @param nested whether to perform a nested group search or not.
     * @return the list of groups that a user is a member of.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public List<GroupDTO> getGroups(String tenant, String name, String domain, boolean nested) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        String upn = UPNUtil.buildUPN(name, domain);
        Map<String, Object> params = buildParameters("nested", String.valueOf(nested));
        URI uri = buildURI(parent.getHostRetriever(), USER_NAME_GROUPS_POST_URI, tenant, upn, params);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken());
        return executeAndReturnList(parent.getClient(), post, GroupDTO.class);
    }

}
